### ISSUE TITLE
"kubectl delete pod" doesn't kill pod's process immediately when specifying --grace-period=0

### ISSUE THREAD

**Issue by YuikoTakada** [kubernetes/kubernetes#120449]
### What happened?

As far as we read the help of "kubectl delete", it looks like that --grace-period=0 immediately kills Pod's processes.
```
    --force=false:
        If true, immediately remove resources from API and bypass graceful deletion. Note that immediate deletion of
        some resources may result in inconsistency or data loss and requires confirmation.

    --grace-period=-1:
        Period of time in seconds given to the resource to terminate gracefully. Ignored if negative. Set to 1 for
        immediate shutdown. Can only be set to 0 when --force is true (force deletion).
```

In fact, processes gets killed if we set --grace-period=0 in ver1.18.
But it seems that the behehavior got changed between ver1.18 and ver1.27.

- ver1.18 ... grace-period=0 killes Pod's processes immediately
- ver1.27 ... grace-period=0 sends SIGTERM to Pod's processes first, then wait for Pod's terminationGracePeriodSeconds before sending SIGKILL. As a result, when I execute `kubectl delete pod grace-period=0 --force` , message "pod pod-name force deleted" is shown, but pod's process is not killed immediately. It lives during `terminationGracePeriodSeconds` .

### What did you expect to happen?

when execute "kubectl delete pod --force --grace-period=0" , process will have been deleted immediately unless wating during `terminationGracePeriodSeconds`

### How can we reproduce it (as minimally and precisely as possible)?

nginx-pod.yaml:
```
apiVersion: v1
kind: Pod
metadata:
  name: nginx
spec:
  terminationGracePeriodSeconds: 300
  containers:
  - name: nginx
    image: nginx:1.14.2
    command: [ 'sleep', '12345']
    ports:
    - containerPort: 80
```

```
$ kubectl apply -f nginx-pod.yaml
pod/nginx created

$ kubectl get pods
NAME    READY   STATUS    RESTARTS   AGE
nginx   1/1     Running   0          10s

$ date; kubectl delete pod nginx  --force --grace-period=0
Wed Aug  2 14:52:15 JST 2023
Warning: Immediate deletion does not wait for confirmation that the running resource has been terminated. The resource may continue to run on the cluster indefinitely.
pod "nginx " force deleted

$ while true; do date; ps -ef | grep 12345; sleep 1; done
Wed Sep  6 17:21:25 JST 2023
root      896616  896559  0 17:20 ?        00:00:00 sleep 12345
root      897253  897201  0 17:21 ?        00:00:00 sleep 12345
yuiko     897353  897155  0 17:21 pts/3    00:00:00 grep 12345
Wed Sep  6 17:21:26 JST 2023
root      896616  896559  0 17:20 ?        00:00:00 sleep 12345
root      897253  897201  0 17:21 ?        00:00:00 sleep 12345
yuiko     897368  897155  0 17:21 pts/3    00:00:00 grep 12345
...
Wed Sep  6 17:26:25 JST 2023
root      897253  897201  0 17:21 ?        00:00:00 sleep 12345
yuiko     899845  897155  0 17:26 pts/3    00:00:00 grep 12345
Wed Sep  6 17:26:26 JST 2023
yuiko     899900  897155  0 17:26 pts/3    00:00:00 grep 12345
Wed Sep  6 17:26:27 JST 2023 17:26:27 JST
-> process has been deleted after 300sec
```

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"27", GitVersion:"v1.27.4", GitCommit:"fa3d7990104d7c1f16943a67f11b154b71f6a132", GitTreeState:"clean", BuildDate:"2023-07-19T12:20:54Z", GoVersion:"go1.20.6", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v5.0.1
Server Version: version.Info{Major:"1", Minor:"28+", GitVersion:"v1.28.1-19+16bd870d488ccb", GitCommit:"16bd870d488ccb24dbe3a4c0cc6afb28896b67e2", GitTreeState:"clean", BuildDate:"2023-09-12T11:10:44Z", GoVersion:"go1.20.7", Compiler:"gc", Platform:"linux/amd64"}
```

</details>

### Cloud provider

<details>
kind 
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>
kind
</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by tamilselvan1102** [comment]
As per the [document ](https://kubernetes.io/docs/tasks/run-application/force-delete-stateful-set-pod/#force-deletion), the pod may stuck in Unknown state for some time while doing pod deletion

**Comment by SergeyKanzhelev** [comment]
@YuikoTakada I remember there was a fix for this in 1.28. Can you please add tests to your PR to prove the scenarios is not working.

/triage needs-information

**Comment by YuikoTakada** [comment]
@SergeyKanzhelev Thank you. I'll test with v1.28.

**Comment by tuibeovince** [comment]
Just for supplementary information; the grace period recalculation started in v1.22. Since then, the logic for changing the grace period values weren't changed until 1.28 (current released version).

**Comment by JCMais** [comment]
is there any workaround for this?

**Comment by YuikoTakada** [comment]
No, There is no workaround currently.

**Comment by smarterclayton** [comment]
The intended behavior here is that when the kubelet observes a pod has been force deleted, that it sends a SIGTERM to container processes as soon as it can, and then should force stop the processes with SIGKILL within 2s to give the pod an opportunity to shutdown cleanly.

Because a force deletion is the removal of a pod without waiting for clean shutdown, the kubelet handles that termination as part of its exceptional cleanup loop, which might take an extra few seconds to react.

The kubelet must never stop a container immediately with SIGKILL, and may only send that signal after attempting a SIGTERM.

**Comment by YuikoTakada** [comment]
@smarterclayton Thank you for your comments. In latest version, `SIGTERM` is sent immediately when `kubectl delete pod --force --grace-period=0` is executed. Because when execute `kubectl get pods` immediately after `kubectl delete pod --force --grace-period=0` , `No resources found in default namespace.` message is shown. But process is still living during `terminationGracePeriodSeconds` (default: 300). 
This behavior is different from ideal behavior that  you said " it sends a SIGTERM to container processes as soon as it can, and then should force stop the processes with SIGKILL within 2s"

**Comment by toyamagu-2021** [comment]
Is this issue fixed by #124063 ?

**Comment by haircommander** [comment]
yeah this looks like it would have been fixed by https://github.com/kubernetes/kubernetes/pull/124063
I think we can
/close
this
